File: src/ros2_utils_py/ros2_utils_py/service_utils.py
import time
import threading
import concurrent.futures
import rclpy
from rclpy.node import Node
from rclpy.client import Client
from rclpy.task import Future
from rclpy.logging import get_logger
from msg_format.srv import ProcessService


def call_service(
        client: Client,
        logger,
        action: str,
        timeout: float = 6.0) -> bool:

    service_name = client.srv_name

    # Wait for service to be available
    if not client.wait_for_service(timeout_sec=timeout):
        logger.error(f"{service_name} service not available after timeout")
        return False
    
    # Create request
    request = ProcessService.Request()
    request.action = action
    
    # Send request
    future = client.call_async(request)
    
    ret = rclpy.spin_until_future_complete(client, future, timeout_sec=timeout)

    if ret == rclpy.FutureReturnCode.SUCCESS:
        logger.info(f"Service {service_name} call succeeded")
        return True
    else:
        logger.error(f"Service {service_name} call timed out")
        return False

def get_command(command: str, device_id: str) -> str:
    """
    Parse command to extract device_specific command.
    
    Args:
        command: The full command string
        device_id: The device ID to look for
        
    Returns:
        str: Extracted command or status string
    """
    if command == "test" or command == "init":
        return command
    
    pos = command.find(device_id)
    if pos == -1:  # -1 is Python's equivalent to std::string::npos
        return "none"
    
    # Get the position of the underscore
    underscore_pos = pos + len(device_id)
    if underscore_pos >= len(command) or command[underscore_pos] != '_':
        return "error"
    
    # Get the command
    space_pos = command.find(' ', underscore_pos + 1)
    if space_pos == -1:
        device_command = command[pos:]
    else:
        device_command = command[pos:space_pos]
    
    # Check if the action is empty
    if not device_command:
        return "error"
    
    return device_command

File: src/ros2_utils/src/service_utils.cpp
#include "ros2_utils/service_utils.hpp"

namespace service_utils {

bool call_service(
    rclcpp::Client<msg_format::srv::ProcessService>::SharedPtr client,
    rclcpp::Logger logger,
    const std::string& action,
    const std::chrono::seconds& timeout)
{
    const std::string service_name = client->get_service_name();
    if (service_name.empty()) {
        RCLCPP_ERROR(logger, "Client has no service name");
        return false;
    }

    if (!client->wait_for_service(timeout)) {
        RCLCPP_ERROR(logger, "%s service not available after %ld seconds", 
                     service_name.c_str(), timeout.count());
        return false;
    }

    auto request = std::make_shared<msg_format::srv::ProcessService::Request>();
    request->action = action;

    auto future = client->async_send_request(request);

    auto ret = rclcpp::spin_until_future_complete(
        rclcpp::Node::make_shared("temp_node"), future, timeout);

    if (ret == rclcpp::FutureReturnCode::SUCCESS) {
        RCLCPP_INFO(logger, "status: %s", future.get()->result.c_str());
        return true;
    } else {
        RCLCPP_ERROR(logger, "Failed to call service %s", service_name.c_str());
        return false;
    }
}

std::string get_command(
    std::string command,
    const std::string device_id) {
    
    if (command == "test" || command == "init"){
        return command;
    }

    std::string::size_type pos = command.find(device_id);
    if (pos == std::string::npos) {
        return "none";
    }

    // get the position of the underscore
    std::string::size_type underscorePos = pos + device_id.length();
    if (underscorePos >= command.length() || command[underscorePos] != '_') {
        return "error";
    }
    // get the command
    std::string::size_type spacePos = command.find(' ', underscorePos + 1);
    std::string device_command = command.substr(pos, spacePos - pos);

    // check if the action is empty
    if (device_command.empty()) {
        return "error";
    }
    
    return device_command;
}

std::vector<std::string> split_string(const std::string& input, char delimiter) {
    std::vector<std::string> tokens;
    std::stringstream ss(input);
    std::string token;
    
    while (std::getline(ss, token, delimiter)) {
        tokens.push_back(token);
    }
    
    return tokens;
}

} // namespace service_utils